ADBDEV-4907-44: Avoid passing nulls to memcpy

RekGRpth opened the pull request:
Avoid passing nulls to memcpy

The appendonly_beginrangescan_internal function can set the
scan->aos_key pointer to NULL, which can then be passed to
memcpy as a destination in the initscan function.

I added two checks to avoid passing a null destination or
a zero length to the memcpy function.

RekGRpth committed:
Avoid passing nulls to memcpy

The appendonly_beginrangescan_internal function can set the
scan->aos_key pointer to NULL, which can then be passed to
memcpy as a destination in the initscan function.

I added two checks to avoid passing a null destination or
a zero length to the memcpy function.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/63315

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1034461

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1034459

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1034462

bandetto commented on src/backend/access/appendonly/appendonlyam.c:
```diff
@@ -148,7 +148,7 @@ initscan(AppendOnlyScanDesc scan, ScanKey key)
 	 * copy the scan key, if appropriate
 	 */
-	if (key != NULL)
+	if (key != NULL && scan->aos_key != NULL && scan->aos_nkeys > 0)
```
```suggestion
	if (key != NULL && scan->aos_nkeys > 0)
```
Maybe remove extra check because the only place where `scan->aos_nkeys = nkeys` is set is at https://github.com/arenadata/gpdb/blob/3b5aa3e9d5141832d94172e20e41a75770c07f34/src/backend/access/appendonly/appendonlyam.c#L1590 and then the count is checked at https://github.com/arenadata/gpdb/blob/3b5aa3e9d5141832d94172e20e41a75770c07f34/src/backend/access/appendonly/appendonlyam.c#L1639-L1642?

RekGRpth replied:
> the only place where `scan->aos_nkeys = nkeys` is set is at

how about appendonly_rescan?

bandetto replied:
I didn't find any other places where `aos_nkeys` is changed.

RekGRpth replied:
why not leave
```suggestion
	if (key != NULL && scan->aos_key != NULL)
```
?

bandetto replied:
I think the condition with `aos_nkeys` shows the expected behavior better. Maybe add an assert before in the condition to silence the static analyzer? I'm okay with the original change as well, but it may confuse future maintainers :(

RekGRpth replied:
I think the NULL check is to prevent UB for memcpy, and the length check is to prevent an unnecessary call.

RekGRpth replied:
> an assert

does not

> silence the static analyzer

bandetto commented:
> I added two checks to avoid passing a null **source** or a zero
length **source** to the memcpy function.

Change "source" to "destination"

RekGRpth commented:
> > I added two checks to avoid passing a null **source** or a zero
> > length **source** to the memcpy function.
> 
> Change "source" to "destination"

changed
